File: backend/services/scoring_engine.py
def combine_scores(rule_result, ai_result):
    """
    Combines rule-based and AI scores, and normalizes all violations.
    """
    def normalize_violation(v):
        if isinstance(v, str):
            return {
                "id": f"MSG_{abs(hash(v)) % 10000}",
                "rule": "System Message",
                "message": v,
                "severity": "MEDIUM",
                "field": None
            }
        
        # Ensure severity is uppercase and valid
        severity = str(v.get("severity", "MEDIUM")).upper()
        if severity not in ["CRITICAL", "HIGH", "MEDIUM", "LOW"]:
            severity = "MEDIUM"
            
        return {
            "id": v.get("id", f"UNK_{abs(hash(v.get('message', ''))) % 10000}"),
            "rule": v.get("rule", "Compliance Rule"),
            "message": v.get("message", "Rule violation detected"),
            "severity": severity,
            "field": v.get("field")
        }

    final_score = int((rule_result["rule_score"] * 0.5) + (ai_result["ai_score"] * 0.5))

    if final_score >= 80:
        risk = "Low"
    elif final_score >= 50:
        risk = "Medium"
    else:
        risk = "High"

    # Normalize all violations from both sources
    raw_violations = rule_result.get("violations", []) + ai_result.get("ai_violations", [])
    normalized_violations = [normalize_violation(v) for v in raw_violations]

    return {
        "score": final_score,
        "rule_score": rule_result["rule_score"],
        "risk": risk,
        "violations": normalized_violations,
        "passed_rules": rule_result["passed_rules"],
        "total_rules": rule_result["total_rules"],
        "ai_analysis_status": ai_result.get("ai_status", "Success")
    }

File: node_backend/services/scoringEngine.js
/**
 * Scoring Engine — equivalent to services/scoring_engine.py
 * Combines rule-based and AI scores, normalizes violations.
 */

export function normalizeViolation(v) {
  if (typeof v === 'string') {
    return {
      id: `MSG_${Math.abs(hashStr(v)) % 10000}`,
      rule: 'System Message',
      message: v,
      severity: 'MEDIUM',
      field: null,
    };
  }

  let severity = String(v.severity || 'MEDIUM').toUpperCase();
  if (!['CRITICAL', 'HIGH', 'MEDIUM', 'LOW', 'INFO'].includes(severity)) {
    severity = 'MEDIUM';
  }

  return {
    id: v.id || `UNK_${Math.abs(hashStr(v.message || '')) % 10000}`,
    rule: v.rule || 'Compliance Rule',
    message: v.message || 'Rule violation detected',
    severity,
    field: v.field || null,
  };
}

function hashStr(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash |= 0; // Convert to 32bit integer
  }
  return hash;
}

export function combineScores(ruleResult, aiResult) {
  const finalScore = Math.round((ruleResult.rule_score * 0.5) + (aiResult.ai_score * 0.5));

  let risk;
  if (finalScore >= 80) risk = 'Low';
  else if (finalScore >= 50) risk = 'Medium';
  else risk = 'High';

  // Normalize all violations from both sources
  const rawViolations = [
    ...(ruleResult.violations || []),
    ...(aiResult.ai_violations || []),
  ];
  const normalizedViolations = rawViolations.map(normalizeViolation);

  return {
    score: finalScore,
    rule_score: ruleResult.rule_score,
    ai_score: aiResult.ai_score,
    risk,
    violations: normalizedViolations,
    passed_rules: ruleResult.passed_rules,
    total_rules: ruleResult.total_rules,
    ai_analysis_status: aiResult.ai_status || 'Success',
    deep_scrape_available: ruleResult.deep_scrape_available || true,
  };
}

export default { combineScores, normalizeViolation };
